sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Redistribute.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms;
+
+import java.util.concurrent.ThreadLocalRandom;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.util.IdentityWindowFn;
+import org.apache.beam.sdk.util.Reshuffle;
+import org.apache.beam.sdk.util.ReshuffleTrigger;
+import org.apache.beam.sdk.util.ValueWithRecordId;
+import org.apache.beam.sdk.util.WindowingStrategy;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Duration;
+
+/**
+ * {@link PTransform Transforms} that return a {@link PCollection} equivalent to their input, but
+ * which is distributed differently, in order to provide additional operational properties. See
+ * {@link #byKey} and {@link #arbitrarily}.
+ */
+public class Redistribute {
+  /**
+   * Redistributes the collection by key, providing some of the side effects of a {@link
+   * GroupByKey}, in particular: prevents fusion of the surrounding transforms, checkpointing and
+   * deduplication by id (see {@link ValueWithRecordId}), and ensures that elements with the same
+   * key are processed by the same worker.
```

Comment:
It would be good to describe this in terms of semantics as well, or primarily. Though it is the identity on the contents of the `PCollection`, it is not the identity on the meta-properties of the `PCollection`. Specifically, it makes it "keyed".

This is a strictly abstract property, named only to give intuition (and because pretty much all runners implement it the same way :-). It means that you can enable specific capabilities on the `PCollection` such as keyed state. Other manifestations of this that one can imagine a related method on `Source` (along the lines of [`producesSortedKeys()`](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedSource.java#L76)) that indicates an initial splitting so that anything fused with the source already has this "keyed" property, and `PreservesKeys` property that a `DoFn<KV, KV>` might have indicating that you can still use keyed state on its output.

Certainly, we should also include the operational description as well, as an example. Users will most easily understand it exactly as written. But, for example, the `DirectRunner` doesn't have workers and this can be a no-op. One can imagine actual distributed runners that have an architecture that makes this property irrelevant (for better or for worse).